### DISCUSSION TITLE
Best practices for mixed C and C++

### DISCUSSION THREAD

**Discussion by lb90** [mesonbuild/meson#10106] [Q&A] [answered]
Hello,

I'm working on adding DirectWrite support to Pango 2.0. Pango is written in C, but in order to use DirectWrite as of today we are forced to use C++, because DirectWrite headers are incomplete with regards to C interfacing ([1], [2], [3]).

I'd like to ask two questions in particular:

1. Is it completely supported by Meson to specify mixed `.c` and `.cpp` files when creating a target? (static libraries, shared libraries or executables)
2. Pango currently calls `cc = meson.get_compiler('c')` at the top of the `meson.build` script. Does the usage of `cc` also affect the C++ compiler? Should I instead add `cxx = meson.get_compiler('cpp')` and mirror all usages of `cc` to `cxx`?

With kind regards,
Thank you!

1. https://devblogs.microsoft.com/oldnewthing/20220113-00/?p=106152
2. https://sourceforge.net/p/mingw-w64/mailman/mingw-w64-public/thread/CAMq1ado5MOQN%3D0xXsHY9f%3DKBHpTHa4W%2Bt-kVkwys5s8k1V5nBQ%40mail.gmail.com/#msg37417555
3. https://sourceforge.net/p/mingw-w64/mailman/mingw-w64-public/thread/aff7dfc0-c5cb-0fd8-5c80-9887a49ebb34%40ycbcr.xyz/#msg37616115

**Comment by eli-schwartz** [comment, marked as answer]
- meson supports adding both C and C++ sources to the same library or executable, and will by default use the C++ driver to link with. There is a `link_language` kwarg to override this.
- both the cc and cxx compilers will exist (be searched for, declared as languages for files each needs to compile). The get_compiler() function merely gives you a handle to that compiler. In many cases, it doesn't matter which one you use since e.g. `find_library()` finds the same libraries in either case and C++ shared libraries are technically linkable by the C compiler because they might provide a C ABI (but you can't really know until you try). Depending on your use case, though, you absolutely do need to use the right compiler depending on what you are doing. For example, `cc.compiles()` will check if the C compiler can compile some code, and `cxx.compiles()` will check if the C++ compiler can compile some code, and obviously these are different because that's why Meson defines two languages for them to begin with. Since you can use C headers from C++, e.g. has_function or has_header(_symbol)? will probably work with either one, as long as you're only checking C stuff.

**Reply by lb90** [reply]
Thanks for the great answer, @eli-schwartz!

I have checked all of the `cc` usages in Pango. For most of them everything is really obvious, nothing to say. Instead I have doubts regarding:

* `add_project_arguments ('-foo ...', language: 'c')`. I should change the `language` arg to `['c', 'cpp']`, right?
* When specifying the `c_flags` argument in a build target, I should also set the `cpp_flags`, otherwise the cpp compiler won't pick up the flags, is that correct?

By the way, it'd be really great to have a dedicated FAQ section on mesonbuild.com 😉

**Reply by xclaesse** [reply]
Yes if you mix C/C++ sources that need the same flags you have to define them in both c_flags and cpp_flags and add_project_arguments().

> By the way, it'd be really great to have a dedicated FAQ section on mesonbuild.com wink

https://mesonbuild.com/FAQ.html

Feel free to open PR to extend the doc :)

**Reply by xclaesse** [reply]
Related: https://mesonbuild.com/FAQ.html#why-dont-the-arguments-passed-to-add_project_link_arguments-affect-anything

**Reply by eli-schwartz** [reply]
In theory you might have flags which are only used by the C sources and not the C++  sources, but... probably not.

So yeah, extend the language to apply it for both.

**Comment by afud** [comment]
@eli-schwartz 
Hello. 

I have a mixed 'c' and 'c++' project, and add all c++ files and c files in the same sources to generate a executable file.  Can I 
 use 'link_language' to avoid error of 'No host machine compiler for xxx.c'? What should I do?

Best wishes
Thanks

**Reply by eli-schwartz** [reply]
No, the link language only controls which compiler is used to link the object files together.

The object files themselves are still created by the correct compiler for each language. Any files with a C file extension require a C compiler, any files with a C++ file extension require a C++ compiler.

If you are getting that error, it sounds like in the `project('myproject', 'c', 'cpp')` you actually *don't* have the C language declared in the list of languages your project uses. So meson doesn't *look* for a C compiler, and thus doesn't have it available when it finds sources named `*.c`.

So if you fix this teeny oversight things should start working, most likely.

**Reply by afud** [reply]
Yes, I'm missing the discription of this.  When adding the 'c' language specifiction, it working. thanks a lot.
![image](https://user-images.githubusercontent.com/49606633/165876840-5b72b9d0-c373-4b7a-988d-646ddf22ee3c.png)  

@eli-schwartz 
anther tiny question about the output name of a library or module, which starts with 'lib' automaticly, is this design for standardization?

Best wishes

**Reply by eli-schwartz** [reply]
On Windows it doesn't much matter, but it's required on Unix in order to work correctly with compilers. For consistency, it starts with "lib" on Windows as well.

**Reply by afud** [reply]
I see, thank you.
